### ISSUE TITLE
Match GPX track to road

### ISSUE THREAD

**Issue by scaidermern** [osmandapp/Osmand#4065]
It would be really nice if OsmAnd could try to match a GPX track to its road data by performing some kind of map matching.

This is an example of navigating along a GPX track with only sparse track points:

![photo_2017-07-07_17-09-20](https://user-images.githubusercontent.com/1040784/27964069-471f7848-6337-11e7-888c-1ae34bb587ce.jpg)

In this case the current visualization is still usable. However there are cases where it becomes unclear which road to take, e.g. in [this situation](http://www.openstreetmap.org/#map=18/47.30022/12.79184):
![photo_2017-07-07_17-09-15](https://user-images.githubusercontent.com/1040784/27964362-51c8f520-6338-11e7-918d-50883abc0736.jpg)

Imagine driving towards this intersection. At first it looks like the user has to keep on the right side of the road (marked with the green casing) due to the exceptional unfavorable position of the GPX points. However the correct decision is to keep on the left road and enter the bridge (marked with the cyan casing).

Also, the current GPX track navigation misses many useful features, like

- maxspeed information (this gets displayed from time to time, but mostly not)
- roundabout exit hints
- lane information
- other useful hints like warnings for pedestrian crossings, railroad crossings, speed cameras, stop signs and so on

Interestingly a T\*mT\*m Rider is able to perform this map matching and is exceptionally fast at it. For a 2000 km GPX track it needs less than a minute. At least this is what the user experiences but maybe the matching is performed on the fly while navigating instead of immediately.

tl;dr: This would be a really nice feature :)

Related to #3993.

**Comment by rindlerblabla** [comment]
Related to #3908 as well?

**Comment by scaidermern** [comment]
> Related to #3908 as well?

Oh, yes this seems to be a duplicate of #3908. I missed that issue.

**Comment by kreuzschnabel** [comment]
If this feature is built in, make it switchable by any means. Road-matched tracks are useless for mapping purpose because they only double up the data which is already there.

**Comment by sonora** [comment]
kreuzschnabel is right. All we may have to do here if we want to do this at all is provide an option to navigate along a GPX track, like we have it already, but with a snap-to-road matching of the points contained in the GPX before calculating a conventional OsmAnd offline route along them.

**Comment by scaidermern** [comment]
Yes, this would be perfect.

**Comment by vshcherb** [comment]
Duplicate #3908
